dev-tools/releng/src/job.rs

```diff
@@ -0,0 +1,284 @@
+            .try_collect::<()>()
+            .await
+    }
+}
+
+impl Job {
+    async fn run(self) -> Result<()> {
+        let result: Result<(), RecvError> = self
+            .wait_for
+            .into_iter()
+            .collect::<FuturesUnordered<_>>()
+            .try_collect::<()>()
+            .await;
+        result.map_err(|_| anyhow!("dependency failed"))?;
+
+        self.future.await?;
+        for sender in self.notify {
+            sender.send(()).ok();
+        }
+        Ok(())
+    }
+}
+
+impl<'a> Selector<'a> {
+    #[track_caller]
+    pub(crate) fn after(self, other: impl AsRef<str>) -> Self {
+        let (sender, receiver) = oneshot::channel();
+        self.jobs
+            .map
+            .get_mut(&self.name)
+            .expect("invalid job name")
+            .wait_for
+            .push(receiver);
+        self.jobs
+            .map
+            .get_mut(other.as_ref())
+            .expect("invalid job name")
+            .notify
+            .push(sender);
+        self
+    }
+}
+
+macro_rules! info_or_error {
+    ($logger:expr, $result:expr, $($tt:tt)*) => {
+        if $result.is_ok() {
+            ::slog::info!($logger, $($tt)*);
+        } else {
+            ::slog::error!($logger, $($tt)*);
+        }
+    };
+}
+
+async fn run_job<F>(
+    logger: Logger,
+    permits: Arc<Semaphore>,
+    name: String,
+    future: F,
+) -> Result<()>
+where
+    F: Future<Output = Result<()>> + 'static,
+{
+    if !PERMIT_NOT_REQUIRED.contains(&name.as_str()) {
+        let _ = permits.acquire_owned().await?;
+    }
+
+    info!(logger, "[{}] running task", name);
+    let start = Instant::now();
+    let result = future.await;
+    let duration = Instant::now().saturating_duration_since(start);
+    info_or_error!(
+        logger,
+        result,
+        "[{}] task {} ({:?})",
+        name,
+        if result.is_ok() { "succeeded" } else { "failed" },
+        duration
+    );
+    result
+}
+
+async fn spawn_with_output(
+    mut command: Command,
+    logger: Logger,
+    permits: Arc<Semaphore>,
+    name: String,
+    log_path: Utf8PathBuf,
+) -> Result<()> {
+    let _ = permits.acquire_owned().await?;
+
+    let log_file_1 = File::create(log_path).await?;
+    let log_file_2 = log_file_1.try_clone().await?;
+
+    info!(logger, "[{}] running: {}", name, command.to_string());
+    let start = Instant::now();
+    let mut child = command
+        .kill_on_drop(true)
+        .stdin(Stdio::null())
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .spawn()
+        .with_context(|| format!("failed to exec `{}`", command.to_string()))?;
+
+    let stdout = reader(
+        &name,
+        child.stdout.take().unwrap(),
+        tokio::io::stdout(),
+        log_file_1,
+    );
+    let stderr = reader(
+        &name,
+        child.stderr.take().unwrap(),
+        tokio::io::stderr(),
+        log_file_2,
+    );
+    match tokio::try_join!(child.wait(), stdout, stderr) {
```

Comment:
This will cancel immediately if any of the operations fail. Is that desirable?

* If so, then can this be documented in a comment?
* If not, then https://docs.rs/cancel-safe-futures/0.1.5/cancel_safe_futures/macro.join_then_try.html exists.